src/ui/public/share/directives/share.js

```diff
@@ -1,18 +1,145 @@
+
 import uiModules from 'ui/modules';
 import shareTemplate from 'ui/share/views/share.html';
 const app = uiModules.get('kibana');
 
-app.directive('share', function () {
+app.directive('share', function (Private) {
+  const getUnhashableStates = Private(getUnhashableStatesProvider);
+  const urlShortener = Private(urlShortenerProvider);
+
   return {
     restrict: 'E',
     scope: {
       objectType: '@',
       objectId: '@',
-      setAllowEmbed: '&?allowEmbed'
+      allowEmbed: '@',
     },
     template: shareTemplate,
-    controller: function ($scope) {
-      $scope.allowEmbed = $scope.setAllowEmbed ? $scope.setAllowEmbed() : true;
+    controllerAs: 'share',
+    controller: function ($scope, $location, globalState) {
+      if ($scope.allowEmbed !== 'false' && $scope.allowEmbed !== undefined) {
+        throw new Error('allowEmbed must be "false" or undefined');
+      }
+
+      // Default to allowing an embedded IFRAME, unless it's explicitly set to false.
+      this.allowEmbed = $scope.allowEmbed === 'false' ? false : true;
+      this.objectType = $scope.objectType;
+
+      function getOriginalUrl() {
+        // If there is no objectId, then it isn't saved, so it has no original URL.
+        if ($scope.objectId === undefined || $scope.objectId === '') {
+          return;
+        }
+
+        // Remove the query string.
+        const url = $location.absUrl();
+        return url.split('?')[0];
+      };
+
+      function getSnapshotUrl() {
+        const url = $location.absUrl();
+        // Replace hashes with original RISON values.
+        const urlWithStates = unhashUrl(url, getUnhashableStates());
+        return urlWithStates;
+      }
+
+      this.makeUrlEmbeddable = url => {
+        const embedQueryParam = '?embed=true';
+        const urlHasQueryString = url.indexOf('?') !== -1;
+        if (urlHasQueryString) {
+          return url.replace('?', `${embedQueryParam}&`);
+        }
+        return `${url}${embedQueryParam}`;
+      };
+
+      this.makeIframeTag = url => {
+        if (!url) return;
+
+        const embeddableUrl = this.makeUrlEmbeddable(url);
+        return `<iframe src="${embeddableUrl}" height="600" width="800"></iframe>`;
+      };
+
+      this.urls = {
+        original: undefined,
+        snapshot: undefined,
+        shortSnapshot: undefined,
+        shortSnapshotIframe: undefined,
+      };
+
+      // By using flags, we can disable the links for generating short URLs once they've been,
+      // clicked. This protects against the user sending multiple requests for short URLs, creating
+      // race conditions.
+      this.urlFlags = {
+        shortSnapshot: false,
+        shortSnapshotIframe: false,
+      };
+
+      const updateUrls = () => {
+        this.urls = {
+          original: getOriginalUrl(),
+          snapshot: getSnapshotUrl(),
+        };
+
+        // Whenever the URL changes, we need to re-generate any short URLs that are displayed.
```

Comment:
I think this is a better user experience, but I'm worried that it could create a bunch of never-used short urls (which in turn are elasticsearch docs)